File: micropython/conway.py
# Python implementation of Conway's Life
# https://en.wikipedia.org/wiki/Conway%27s_Game_of_Life

# A matrix of cells, which contain a running instance
# of Conway's Life.
#
# The neighbourhood (n) around a cell (c) is consulted to
# see if the cell should maintain or change state
#
#  nnn
#  ncn
#  nnn

import random
import cube_bit

live = 1
dead = 0


def random_state():
    if random.randrange(2) == 0:
        return dead
    else:
        return live


def make_world(size):
    return [random_state() for _ in range(size)]


def neighbour_weight(world, cell):
    count = 0
    for i in range(8):
        world_offset = cube_bit.neighbour_of(cell, i)
        count += world[world_offset]
    
    return count


def next_state(state, alive_neighbours):
    new_state = dead
    
    if state == live and (alive_neighbours == 2 or alive_neighbours == 3):
        new_state = live
    elif state == dead and alive_neighbours == 3:
        new_state = live

    return new_state


def next_generation(world):
    next_gen = []
    for cell in range(len(world)):
        nw = neighbour_weight(world, cell)
        next_gen.append(next_state(world[cell], nw))
    
    return next_gen

File: C/conway.c
/*
 * An implementation of Conway's Life
 */

#include "conway.h"
#include "cube_bit.h"

#include <stdlib.h>

static uint8_t random_state() {
    unsigned int half_way = (((unsigned int)RAND_MAX) + 1) / 2;
    return rand() < half_way ? 1 : 0;
}

void init_world(uint8_t* world, size_t count) {
    for (int i = 0; i < count; i++) {
        world[i] = random_state();
    }
}

static uint8_t neighbour_weight(uint8_t* world, uint8_t cell) {
    uint8_t count = 0;
    for (int i = 0; i < 8; i++) {
        uint8_t n = neighbour_of(cell, i);
        count += world[n];
    }
    return count;
}

static uint8_t next_state(uint8_t state, uint8_t alive_neighbours) {
    uint8_t new_state = 0;

    if (state == 1 && (alive_neighbours == 2 || alive_neighbours == 3)) {
        new_state = 1;
    }
    else if (state == 0 && alive_neighbours == 3) {
        new_state = 1;
    }
    return new_state;
}

void next_generation(uint8_t* current_world, uint8_t* next_world, size_t count) {
    for (int i = 0; i < count; i++) {
        uint8_t nw = neighbour_weight(current_world, i);
        next_world[i] = next_state(current_world[i], nw);
    }
}